src/qibolab/platforms/abstract.py

```diff
@@ -267,6 +297,46 @@ def create_RX_pulse(self, qubit, start=0, relative_phase=0):
 
         return Pulse(start, qd_duration, qd_amplitude, qd_frequency, relative_phase, qd_shape, qd_channel, qubit=qubit)
 
+    def create_CZ_pulse_sequence(self, qubits, start=0):
+        # Check in the settings if qubits[0]-qubits[1] is a key
+        if f"{qubits[0]}-{qubits[1]}" in self.settings["native_gates"]["two_qubit"]:
+            pulse_sequence_settings = self.settings["native_gates"]["two_qubit"][f"{qubits[0]}-{qubits[1]}"]["CZ"]
+        elif f"{qubits[1]}-{qubits[0]}" in self.settings["native_gates"]["two_qubit"]:
+            pulse_sequence_settings = self.settings["native_gates"]["two_qubit"][f"{qubits[1]}-{qubits[0]}"]["CZ"]
+        else:
+            raise_error(
+                ValueError,
+                f"Calibration for CZ gate between qubits {qubits[0]} and {qubits[1]} not found.",
+            )
+
+        # If settings contains only one pulse dictionary, convert it into a list that can be iterated below
+        if isinstance(pulse_sequence_settings, dict):
+            pulse_sequence_settings = [pulse_sequence_settings]
+
+        from qibolab.pulses import FluxPulse, PulseSequence
+
+        sequence = PulseSequence()
+        sequence.virtual_z_phases = {}
+
+        for pulse_settings in pulse_sequence_settings:
+            if pulse_settings["type"] == "qf":
+                qf_duration = pulse_settings["duration"]
+                qf_amplitude = pulse_settings["amplitude"]
+                qf_shape = pulse_settings["shape"]
+                qubit = pulse_settings["qubit"]
+                qf_channel = self.settings["qubit_channel_map"][qubit][2]
+                sequence.add(
+                    FluxPulse(
+                        start + pulse_settings["relative_start"], qf_duration, qf_amplitude, qf_shape, qf_channel, qubit
+                    )
+                )
+            elif pulse_settings["type"] == "virtual_z":
+                if not pulse_settings["qubit"] in sequence.virtual_z_phases:
+                    sequence.virtual_z_phases[pulse_settings["qubit"]] = 0
+                else:
+                    sequence.virtual_z_phases[pulse_settings["qubit"]] += pulse_settings["phase"]
```

Comment:
This should be
```suggestion
                if not pulse_settings["qubit"] in sequence.virtual_z_phases:
                    sequence.virtual_z_phases[pulse_settings["qubit"]] = 0
                sequence.virtual_z_phases[pulse_settings["qubit"]] += pulse_settings["phase"]
```
otherwise the first virtual-z pulse on each qubit is ignored.

By the way, you can define `sequence.virtual_z_phases` as [`collections.defaultdict(int)`](https://docs.python.org/3/library/collections.html#collections.defaultdict). This is exactly the same as the standard dictionary, but if you try to access an element that does not exist it will automatically initialize it to 0. So it saves you from all these `if` conditions that only check if elements exist here and in a few other places.

Reply:
Thanks @stavros11 for spotting that bug.
Introducing `defaultdict` as suggested the problem is gone:
```python
        for pulse_settings in pulse_sequence_settings:
            if pulse_settings["type"] == "qf":
                qf_duration = pulse_settings["duration"]
                qf_amplitude = pulse_settings["amplitude"]
                qf_shape = pulse_settings["shape"]
                qubit = pulse_settings["qubit"]
                qf_channel = self.settings["qubit_channel_map"][qubit][2]
                sequence.add(
                    FluxPulse(
                        start + pulse_settings["relative_start"], qf_duration, qf_amplitude, qf_shape, qf_channel, qubit
                    )
                )
            elif pulse_settings["type"] == "virtual_z":
                virtual_z_phases[pulse_settings["qubit"]] += pulse_settings["phase"]
```